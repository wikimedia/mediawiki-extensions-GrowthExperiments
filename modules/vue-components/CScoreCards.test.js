jest.mock( './icons.json', () => ( {
	cdxIconEdit: '',
	cdxIconUserTalk: '',
	cdxIconClock: '',
	cdxIconChart: '',
	cdxIconClose: '',
	cdxIconInfo: '',
	cdxIconInfoFilled: '',
} ), { virtual: true } );
const { mount } = require( '@vue/test-utils' );
const ScoreCards = require( './CScoreCards.vue' );
const { useUserImpact } = require( '../ext.growthExperiments.Homepage.Impact/composables/useUserImpact.js' );
const jsonData = require( './__mocks__/serverExportedData.json' );

const renderComponent = ( { props = {}, provide = {} } = {} ) => mount( ScoreCards, {
	props: Object.assign( {
		userName: 'Alice',
		renderThirdPerson: false,
		hasIntl: true,
	}, props ),
	global: {
		provide: Object.assign( {
			RENDER_MODE: 'desktop',
			IMPACT_MAX_EDITS: 1000,
			IMPACT_MAX_THANKS: 1000,
			$log: jest.fn(),
		}, provide ),
		mocks: {
			$filters: {
				convertNumber: jest.fn( ( x ) => `${ x }` ),
			},
		},
	},
} );

describe( 'ScoreCards', () => {
	beforeAll( () => {
		// Moment will use Date.now() when parsing "lastEditTimestamp" with moment().fromNow().
		// The lastEditTimestamp in /__mocks__/serverExportedData.json is for the Dec 14 2022.
		const JS_DATE_A_MONTH_AFTER_LAST_EDIT = new Date( '2023-01-14' );
		jest.useFakeTimers();
		jest.setSystemTime( JS_DATE_A_MONTH_AFTER_LAST_EDIT );
	} );

	afterAll( () => {
		// Undo the forced time we applied earlier, reset to system default.
		jest.setSystemTime( jest.getRealSystemTime() );
		jest.useRealTimers();
	} );
	it( 'renders correctly without data', () => {
		mw.util.getUrl = jest.fn()
			.mockReturnValue( 'http://default.url' )
			.mockReturnValueOnce( 'http://contributions.url' )
			.mockReturnValueOnce( 'http://thanks.url' );
		const wrapper = renderComponent( {
			props: {
				data: null,
			},
		} );
		expect( wrapper.element ).toMatchSnapshot();
	} );
	it( 'renders correctly with data', () => {
		mw.util.getUrl = jest.fn()
			.mockReturnValue( 'http://default.url' )
			.mockReturnValueOnce( 'http://contributions.url' )
			.mockReturnValueOnce( 'http://thanks.url' );
		const wrapper = renderComponent( {
			props: {
				data: useUserImpact( 60, jsonData ).value,
			},
		} );
		expect( wrapper.element ).toMatchSnapshot();
	} );
	it( 'displays the formatted limit in the over-limit message', () => {
		const props = {
			data: {
				receivedThanksCount: 5000,
			},
			hasIntl: true,
		};

		const wrapper = mount( ScoreCards, {
			props,
			global: {
				provide: {
					RENDER_MODE: 'desktop',
					IMPACT_MAX_EDITS: 1000,
					IMPACT_MAX_THANKS: 2000,
				},
				mocks: {
					$i18n: ( key, ...args ) => ( {
						text: () => {
							if ( key === 'growthexperiments-homepage-impact-scores-over-limit' ) {
								return `${ args[ 0 ] }+`;
							}
							return key;
						},
					} ),
					$filters: {
						convertNumber: jest.fn( ( x ) => `${ x }` ),
					},
				},
			},
		} );

		expect( wrapper.vm.receivedThanksCount ).toBe( '2000+' );
	} );
} );
