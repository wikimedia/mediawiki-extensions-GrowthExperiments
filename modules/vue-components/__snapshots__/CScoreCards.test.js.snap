// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScoreCards renders correctly with data 1`] = `
<div
  class="ext-growthExperiments-ScoreCards"
>
  <div
    class="ext-growthExperiments-ScoreCard"
  >
    <div
      class="ext-growthExperiments-ScoreCard__data-display"
    >
      <span
        class="cdx-icon cdx-icon--medium ext-growthExperiments-ScoreCard__data-display__icon"
      >
        <svg
          height="20"
          viewBox="0 0 20 20"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
          xmlns:xlink="http://www.w3.org/1999/xlink"
        >
          <title>
            growthexperiments-homepage-impact-scores-edit-count
          </title>
          <path
            d=""
          />
        </svg>
      </span>
      
      <div
        class="ext-growthExperiments-CText ext-growthExperiments-CText--size-md ext-growthExperiments-CText--weight-bold"
      >
        
        <a
          class="ext-growthExperiments-ScoreCards__link"
          data-link-id="impact-total-edits"
          href="http://contributions.url"
        >
          4
        </a>
        
      </div>
      
    </div>
    <div
      class="ext-growthExperiments-ScoreCard__label"
    >
      <span
        class="ext-growthExperiments-CText ext-growthExperiments-CText--color-subtle ext-growthExperiments-ScoreCard__label__text"
      >
        
        growthexperiments-homepage-impact-scores-edit-count
        
      </span>
      <!--v-if-->
    </div>
  </div>
  <div
    class="ext-growthExperiments-ScoreCard"
  >
    <div
      class="ext-growthExperiments-ScoreCard__data-display"
    >
      <span
        class="cdx-icon cdx-icon--medium ext-growthExperiments-ScoreCard__data-display__icon"
      >
        <svg
          height="20"
          viewBox="0 0 20 20"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
          xmlns:xlink="http://www.w3.org/1999/xlink"
        >
          <title>
            growthexperiments-homepage-impact-scores-thanks-count
          </title>
          <path
            d=""
          />
        </svg>
      </span>
      
      <div
        class="ext-growthExperiments-CText ext-growthExperiments-CText--size-md ext-growthExperiments-CText--weight-bold"
      >
        
        <a
          class="ext-growthExperiments-ScoreCards__link"
          data-link-id="impact-thanks-log"
          href="http://thanks.url"
        >
          0
        </a>
        
      </div>
      
    </div>
    <div
      class="ext-growthExperiments-ScoreCard__label ext-growthExperiments-ScoreCard__label--with-info"
    >
      <span
        class="ext-growthExperiments-CText ext-growthExperiments-CText--color-subtle ext-growthExperiments-ScoreCard__label__text"
      >
        
        growthexperiments-homepage-impact-scores-thanks-count
        
      </span>
      <span>
        <div
          class="ext-growthExperiments-Popover"
        >
          <div>
            <div>
              
              <button
                aria-label="growthexperiments-homepage-impact-scores-thanks-info-label"
                class="cdx-button cdx-button--action-default cdx-button--weight-quiet cdx-button--size-medium cdx-button--icon-only ext-growthExperiments-ScoreCards__info-button"
              >
                
                <span
                  class="cdx-icon cdx-icon--medium"
                >
                  <svg
                    aria-hidden="true"
                    height="20"
                    viewBox="0 0 20 20"
                    width="20"
                    xmlns="http://www.w3.org/2000/svg"
                    xmlns:xlink="http://www.w3.org/1999/xlink"
                  >
                    <!---->
                    <path
                      d=""
                    />
                  </svg>
                </span>
                
              </button>
              
            </div>
            <div
              class="ext-growthExperiments-Popover__popover-container"
            >
              <!--v-if-->
            </div>
          </div>
        </div>
      </span>
    </div>
  </div>
  <div
    class="ext-growthExperiments-ScoreCard"
  >
    <div
      class="ext-growthExperiments-ScoreCard__data-display"
    >
      <span
        class="cdx-icon cdx-icon--medium ext-growthExperiments-ScoreCard__data-display__icon"
      >
        <svg
          height="20"
          viewBox="0 0 20 20"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
          xmlns:xlink="http://www.w3.org/1999/xlink"
        >
          <title>
            growthexperiments-homepage-impact-recent-activity-last-edit-text
          </title>
          <path
            d=""
          />
        </svg>
      </span>
      
      <span
        class="ext-growthExperiments-CText ext-growthExperiments-CText--size-md ext-growthExperiments-CText--weight-bold"
      >
        
        a month ago
        
      </span>
      
    </div>
    <div
      class="ext-growthExperiments-ScoreCard__label"
    >
      <span
        class="ext-growthExperiments-CText ext-growthExperiments-CText--color-subtle ext-growthExperiments-ScoreCard__label__text"
      >
        
        growthexperiments-homepage-impact-recent-activity-last-edit-text
        
      </span>
      <!--v-if-->
    </div>
  </div>
  <div
    class="ext-growthExperiments-ScoreCard"
  >
    <div
      class="ext-growthExperiments-ScoreCard__data-display"
    >
      <span
        class="cdx-icon cdx-icon--medium ext-growthExperiments-ScoreCard__data-display__icon"
      >
        <svg
          height="20"
          viewBox="0 0 20 20"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
          xmlns:xlink="http://www.w3.org/1999/xlink"
        >
          <title>
            growthexperiments-homepage-impact-recent-activity-best-streak-text
          </title>
          <path
            d=""
          />
        </svg>
      </span>
      
      <span
        class="ext-growthExperiments-CText ext-growthExperiments-CText--size-md ext-growthExperiments-CText--weight-bold"
      >
        
        growthexperiments-homepage-impact-recent-activity-streak-count-text:[1]
        
      </span>
      
    </div>
    <div
      class="ext-growthExperiments-ScoreCard__label ext-growthExperiments-ScoreCard__label--with-info"
    >
      <span
        class="ext-growthExperiments-CText ext-growthExperiments-CText--color-subtle ext-growthExperiments-ScoreCard__label__text"
      >
        
        growthexperiments-homepage-impact-recent-activity-best-streak-text
        
      </span>
      <span>
        <div
          class="ext-growthExperiments-Popover"
        >
          <div>
            <div>
              
              <button
                aria-label="growthexperiments-homepage-impact-scores-streak-info-label"
                class="cdx-button cdx-button--action-default cdx-button--weight-quiet cdx-button--size-medium cdx-button--icon-only ext-growthExperiments-ScoreCards__info-button"
              >
                
                <span
                  class="cdx-icon cdx-icon--medium"
                >
                  <svg
                    aria-hidden="true"
                    height="20"
                    viewBox="0 0 20 20"
                    width="20"
                    xmlns="http://www.w3.org/2000/svg"
                    xmlns:xlink="http://www.w3.org/1999/xlink"
                  >
                    <!---->
                    <path
                      d=""
                    />
                  </svg>
                </span>
                
              </button>
              
            </div>
            <div
              class="ext-growthExperiments-Popover__popover-container"
            >
              <!--v-if-->
            </div>
          </div>
        </div>
      </span>
    </div>
  </div>
</div>
`;

exports[`ScoreCards renders correctly without data 1`] = `
<div
  class="ext-growthExperiments-ScoreCards"
>
  <div
    class="ext-growthExperiments-ScoreCard"
  >
    <div
      class="ext-growthExperiments-ScoreCard__data-display"
    >
      <span
        class="cdx-icon cdx-icon--medium ext-growthExperiments-ScoreCard__data-display__icon"
      >
        <svg
          height="20"
          viewBox="0 0 20 20"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
          xmlns:xlink="http://www.w3.org/1999/xlink"
        >
          <title>
            growthexperiments-homepage-impact-scores-edit-count
          </title>
          <path
            d=""
          />
        </svg>
      </span>
      
      <div
        class="ext-growthExperiments-CText ext-growthExperiments-CText--size-md ext-growthExperiments-CText--weight-bold"
      >
        
        <a
          class="ext-growthExperiments-ScoreCards__link"
          data-link-id="impact-total-edits"
          href="http://contributions.url"
        >
          ...
        </a>
        
      </div>
      
    </div>
    <div
      class="ext-growthExperiments-ScoreCard__label"
    >
      <span
        class="ext-growthExperiments-CText ext-growthExperiments-CText--color-subtle ext-growthExperiments-ScoreCard__label__text"
      >
        
        growthexperiments-homepage-impact-scores-edit-count
        
      </span>
      <!--v-if-->
    </div>
  </div>
  <div
    class="ext-growthExperiments-ScoreCard"
  >
    <div
      class="ext-growthExperiments-ScoreCard__data-display"
    >
      <span
        class="cdx-icon cdx-icon--medium ext-growthExperiments-ScoreCard__data-display__icon"
      >
        <svg
          height="20"
          viewBox="0 0 20 20"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
          xmlns:xlink="http://www.w3.org/1999/xlink"
        >
          <title>
            growthexperiments-homepage-impact-scores-thanks-count
          </title>
          <path
            d=""
          />
        </svg>
      </span>
      
      <div
        class="ext-growthExperiments-CText ext-growthExperiments-CText--size-md ext-growthExperiments-CText--weight-bold"
      >
        
        <a
          class="ext-growthExperiments-ScoreCards__link"
          data-link-id="impact-thanks-log"
          href="http://thanks.url"
        >
          ...
        </a>
        
      </div>
      
    </div>
    <div
      class="ext-growthExperiments-ScoreCard__label ext-growthExperiments-ScoreCard__label--with-info"
    >
      <span
        class="ext-growthExperiments-CText ext-growthExperiments-CText--color-subtle ext-growthExperiments-ScoreCard__label__text"
      >
        
        growthexperiments-homepage-impact-scores-thanks-count
        
      </span>
      <span>
        <div
          class="ext-growthExperiments-Popover"
        >
          <div>
            <div>
              
              <button
                aria-label="growthexperiments-homepage-impact-scores-thanks-info-label"
                class="cdx-button cdx-button--action-default cdx-button--weight-quiet cdx-button--size-medium cdx-button--icon-only ext-growthExperiments-ScoreCards__info-button"
              >
                
                <span
                  class="cdx-icon cdx-icon--medium"
                >
                  <svg
                    aria-hidden="true"
                    height="20"
                    viewBox="0 0 20 20"
                    width="20"
                    xmlns="http://www.w3.org/2000/svg"
                    xmlns:xlink="http://www.w3.org/1999/xlink"
                  >
                    <!---->
                    <path
                      d=""
                    />
                  </svg>
                </span>
                
              </button>
              
            </div>
            <div
              class="ext-growthExperiments-Popover__popover-container"
            >
              <!--v-if-->
            </div>
          </div>
        </div>
      </span>
    </div>
  </div>
  <div
    class="ext-growthExperiments-ScoreCard"
  >
    <div
      class="ext-growthExperiments-ScoreCard__data-display"
    >
      <span
        class="cdx-icon cdx-icon--medium ext-growthExperiments-ScoreCard__data-display__icon"
      >
        <svg
          height="20"
          viewBox="0 0 20 20"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
          xmlns:xlink="http://www.w3.org/1999/xlink"
        >
          <title>
            growthexperiments-homepage-impact-recent-activity-last-edit-text
          </title>
          <path
            d=""
          />
        </svg>
      </span>
      
      <span
        class="ext-growthExperiments-CText ext-growthExperiments-CText--size-md ext-growthExperiments-CText--weight-bold"
      >
        
        ...
        
      </span>
      
    </div>
    <div
      class="ext-growthExperiments-ScoreCard__label"
    >
      <span
        class="ext-growthExperiments-CText ext-growthExperiments-CText--color-subtle ext-growthExperiments-ScoreCard__label__text"
      >
        
        growthexperiments-homepage-impact-recent-activity-last-edit-text
        
      </span>
      <!--v-if-->
    </div>
  </div>
  <div
    class="ext-growthExperiments-ScoreCard"
  >
    <div
      class="ext-growthExperiments-ScoreCard__data-display"
    >
      <span
        class="cdx-icon cdx-icon--medium ext-growthExperiments-ScoreCard__data-display__icon"
      >
        <svg
          height="20"
          viewBox="0 0 20 20"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
          xmlns:xlink="http://www.w3.org/1999/xlink"
        >
          <title>
            growthexperiments-homepage-impact-recent-activity-best-streak-text
          </title>
          <path
            d=""
          />
        </svg>
      </span>
      
      <span
        class="ext-growthExperiments-CText ext-growthExperiments-CText--size-md ext-growthExperiments-CText--weight-bold"
      >
        
        ...
        
      </span>
      
    </div>
    <div
      class="ext-growthExperiments-ScoreCard__label ext-growthExperiments-ScoreCard__label--with-info"
    >
      <span
        class="ext-growthExperiments-CText ext-growthExperiments-CText--color-subtle ext-growthExperiments-ScoreCard__label__text"
      >
        
        growthexperiments-homepage-impact-recent-activity-best-streak-text
        
      </span>
      <span>
        <div
          class="ext-growthExperiments-Popover"
        >
          <div>
            <div>
              
              <button
                aria-label="growthexperiments-homepage-impact-scores-streak-info-label"
                class="cdx-button cdx-button--action-default cdx-button--weight-quiet cdx-button--size-medium cdx-button--icon-only ext-growthExperiments-ScoreCards__info-button"
              >
                
                <span
                  class="cdx-icon cdx-icon--medium"
                >
                  <svg
                    aria-hidden="true"
                    height="20"
                    viewBox="0 0 20 20"
                    width="20"
                    xmlns="http://www.w3.org/2000/svg"
                    xmlns:xlink="http://www.w3.org/1999/xlink"
                  >
                    <!---->
                    <path
                      d=""
                    />
                  </svg>
                </span>
                
              </button>
              
            </div>
            <div
              class="ext-growthExperiments-Popover__popover-container"
            >
              <!--v-if-->
            </div>
          </div>
        </div>
      </span>
    </div>
  </div>
</div>
`;
